proxy/http/remap/PluginDso.cc

```diff
@@ -0,0 +1,265 @@
+/** @file
+
+  A class that deals with plugin Dynamic Shared Objects (DSO)
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  @section details Details
+
+  Implements code necessary for Reverse Proxy which mostly consists of
+  general purpose hostname substitution in URLs.
+
+ */
+
+#include "PluginDso.h"
+#ifdef PLUGIN_DSO_TESTS
+#include "unit-tests/plugin_testing_common.h"
+#else
+#include "tscore/Diags.h"
+#endif
+
+PluginDso::PluginDso(const fs::path &configPath, const fs::path &effectivePath, const fs::path &runtimePath)
+  : _configPath(configPath), _effectivePath(effectivePath), _runtimePath(runtimePath)
+{
+}
+
+PluginDso::~PluginDso()
+{
+  std::string error;
+  (void)unload(error);
+}
+
+bool
+PluginDso::load(std::string &error)
+{
+  /* Clear all errors */
+  error.clear();
+  _errorCode.clear();
+  bool result = true;
+
+  if (isLoaded()) {
+    error.append("plugin already loaded");
+    return false;
+  }
+
+  Debug(_tag, "plugin '%s' started loading DSO", _configPath.c_str());
+
+  /* Find plugin DSO looking through the search dirs */
+  if (_effectivePath.empty()) {
+    error.append("empty effective path");
+    result = false;
+  } else {
+    Debug(_tag, "plugin '%s' effective path: %s", _configPath.c_str(), _effectivePath.c_str());
+
+    /* Copy the installed plugin DSO to a runtime directory */
+    std::error_code ec;
+    if (!copy(_effectivePath, _runtimePath, ec)) {
+      std::string temp_error;
```

Comment:
Not worth changing now, but you could use `bwprint` here, e.g.
```
bwprint(error, "Failed to create a copy: {}", ec);
```

Reply:
I see, sure, will leave it like this and consider using the suggested [internal library](https://docs.trafficserver.apache.org/en/8.0.x/developer-guide/internal-libraries/buffer-writer.en.html) in the future